backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/impl/ElasticsearchBackendFactory.java

```diff
@@ -147,10 +152,27 @@ public BackendImplementor create(EventContext eventContext, BackendBuildContext
 		}
 	}
 
-	private MultiTenancyStrategy getMultiTenancyStrategy(ConfigurationPropertySource propertySource) {
-		MultiTenancyStrategyName multiTenancyStrategyName = MULTI_TENANCY_STRATEGY.get( propertySource );
+	private MultiTenancyStrategy getMultiTenancyStrategy(ConfigurationPropertySource propertySource,
+			EventContext eventContext, BackendBuildContext buildContext) {
+		Optional<MultiTenancyStrategyName> multiTenancyStrategyName = MULTI_TENANCY_STRATEGY.get( propertySource );
+		if ( !multiTenancyStrategyName.isPresent() ) {
+			// the default depends on mapping
+			return buildContext.multiTenancyEnabled() ?
+					new DiscriminatorMultiTenancyStrategy() : new NoMultiTenancyStrategy();
+		}
+
+		MultiTenancyStrategyName multiTenancyStrategy = multiTenancyStrategyName.get();
+		// check multiTenancyStrategy mismatch: required by the mapper vs explicitly configured with properties
+		if ( MultiTenancyStrategyName.NONE.equals( multiTenancyStrategy ) &&
+				buildContext.multiTenancyEnabled() ) {
+			throw log.multiTenancyRequiredButExplicitlyDisabledByBackend( eventContext );
+		}
+		if ( MultiTenancyStrategyName.DISCRIMINATOR.equals( multiTenancyStrategy ) &&
+				!buildContext.multiTenancyEnabled() ) {
+			throw log.multiTenancyNotRequiredButExplicitlyEnabledByTheBackend( eventContext );
+		}
```

Comment:
Note you can get even better error messages, with the property name in the error message, by moving this check where you get the property value.

And by the way, you should not need to pass context to your log method, there should be some try/catch higher in the call stack that adds the context automatically.

For instance:

```java
Optional<MultiTenancyStrategyName> multiTenancyStrategyName = MULTI_TENANCY_STRATEGY.getAndTransform( propertySource, name -> {
	boolean enabled = !MultiTenancyStrategyName.NONE.equals( multiTenancyStrategy );
	if ( !enabled && buildContext.multiTenancyEnabled() ) {
		throw log.multiTenancyRequiredButExplicitlyDisabledByBackend();
	}
	else if ( enabled && !buildContext.multiTenancyEnabled() ) {
		throw log.multiTenancyNotRequiredButExplicitlyEnabledByTheBackend();
	}
	return name;
} );
```

Then you'll get an error like:

```
default backend:
  failures:
     - Invalid value for property 'hibernate.search.backend.multitenancy.strategy': 'NONE'. The mapping requires multi-tenancy but no multi-tenancy strategy is set.
```

You can find a similar, though more complex, example in `org.hibernate.search.backend.elasticsearch.impl.ElasticsearchLinkImpl#initVersion`.